### PR TITLE
std: Internalize almost all of `std::rt`

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit does some refactoring to make almost all of the `std::rt` private.
Specifically, the following items are no longer part of its API:
- DEFAULT_ERROR_CODE
- backtrace
- unwind
- args
- at_exit
- cleanup
- heap (this is just alloc::heap)
- min_stack
- util

The module is now tagged as `#[doc(hidden)]` as the only purpose it's serve is
an entry point for the `panic!` macro via the `begin_unwind` and
`begin_unwind_fmt` reexports.

### CODE REVIEW

File: src/liblog/lib.rs [hunk 1]
```diff
@@ -442,19 +440,6 @@ fn init() {
 
         assert!(DIRECTIVES.is_null());
         DIRECTIVES = Box::into_raw(box directives);
-
-        // Schedule the cleanup for the globals for when the runtime exits.
```
Comment (ranma42) [inline src/liblog/lib.rs]: If this block is to be removed, should we also delete https://github.com/alexcrichton/rust/blob/less-rt/src/liblog/lib.rs#L293 and https://github.com/alexcrichton/rust/blob/less-rt/src/liblog/lib.rs#L386 ?
AFAICT `1` will never be assigned to them in the new code.

File: src/libstd/rt.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2013 The Rust Project Developers. See the COPYRIGHT
```
Comment (ranma42) [inline src/libstd/rt.rs]: Is the copyright year intentionally 2013?

File: src/libstd/rt.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        // Store our args if necessary in a squirreled away location
+        sys_common::args::init(argc, argv);
+
+        // And finally, let's run some code!
+        let res = thread::catch_panic(mem::transmute::<_, fn()>(main));
+        sys_common::cleanup();
+        res.is_err()
+    };
+
+    // If the exit code wasn't set, then the try block must have panicked.
```
Comment (retep998) [inline src/libstd/rt.rs]: Comment seems irrelevant since it isn't checking any exit code?

File: src/libstd/rt.rs (hunk 2 above)
Comment (alexcrichton) [inline src/libstd/rt.rs]: Yes this was just moved over from where it was before

File: src/liblog/lib.rs (hunk 1 above)
Comment (alexcrichton) [inline src/liblog/lib.rs]: Yeah, those pieces can be deleted

**Comment by arielb1** [comment]
Is the privacy-ing of `sys::backtrace::write` intentional? That function is really important for anyone that wants to write their own panic handler.

**Comment by brson** [comment]
r=me assuming @arielb1's q's are satisfied

**Comment by alexcrichton** [comment]
@arielb1 it is intentional, yes. This module was never intended to be public and it's just a historical relic that it's still exposed, so I wanted to hide it via privacy ASAP. Currently there's no stable method to write your own panic handler, so it's not a critical use case for continued exposure. External crates can use [support on crates.io](https://crates.io/crates/backtrace) and this can be re-exposed temporarily for internal crates if truly necessary.

@bors: r=brson 53f933b661c9206d29a2414d58a2f106aed0db9b
